From b5b9cd2657eae72ed5718a495fae2f5d0b4bccfb Mon Sep 17 00:00:00 2001
From: Ghennadi Procopciuc <ghennadi.procopciuc@nxp.com>
Date: Thu, 1 Apr 2021 16:53:12 +0300
Subject: [PATCH 05/14] s32g: clk: Check clock state before returning errors

Do not throw errors when receiving a clock reconfiguration
when the clock is running at the requested frequency.

Upstream-Status: Pending

Issue: ALB-6451
Signed-off-by: Ghennadi Procopciuc <Ghennadi.Procopciuc@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 plat/nxp/s32g/s32g_scmi_clk.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/plat/nxp/s32g/s32g_scmi_clk.c b/plat/nxp/s32g/s32g_scmi_clk.c
index cf0029273..86a4c2d0a 100644
--- a/plat/nxp/s32g/s32g_scmi_clk.c
+++ b/plat/nxp/s32g/s32g_scmi_clk.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2020 NXP
+ * Copyright 2020-2021 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -147,6 +147,11 @@ int32_t plat_scmi_clock_set_rate(unsigned int agent_id, unsigned int scmi_id,
 	struct clk_driver *drv;
 	struct clk clk;
 
+	/* Already running at the requested frequency */
+	if (s32gen1_scmi_clk_is_enabled(scmi_id) &&
+	    plat_scmi_clock_get_rate(agent_id, scmi_id) == rate)
+		return SCMI_SUCCESS;
+
 	/**
 	 * Limitation: The rate of a clock cannot be
 	 * changed once it's enabled
-- 
2.25.1

